src/Form/TfaLoginForm.php

```diff
@@ -70,20 +70,22 @@ public function buildForm(array $form, FormStateInterface $form_state) {
    */
   public function submitForm(array &$form, FormStateInterface $form_state) {
     // Similar to tfa_user_login() but not required to force user logout.
+    /** @var $account \Drupal\user\UserInterface */
+    $account = $this->userStorage->load($form_state->get('uid'));
 
-    if ($uid = $form_state->get('uid')) {
-      $account = $this->userStorage->load($uid);
-    }
-    else {
-      $account = user_load_by_name($form_state->get('name'));
-    }
 
-    if ($tfa = $this->tfaManager->getProcess($account)) {
-      if ($account->hasPermission('require tfa') && !$this->tfaManager->loginComplete($account) && !$tfa->ready()) {
+    //GetPlugin
+    //Pass to service functions.
+    /** @var  $tfaValidationPlugin \Drupal\tfa\Plugin\TfaValidationInterface */
+    $tfaValidationPlugin = $this->tfaValidationManager->getInstance(['uid' => $account->id()]);
+
+    //Setup TFA
+    if (isset($tfaValidationPlugin)) {
+      if ($account->hasPermission('require tfa') && !$this->loginComplete($account) && !$this->ready($tfaValidationPlugin)) {
```

Comment:
Hmmmm this is a bit surprising to me that we're doing permission based assignment. We should look down the road as an add-on to make a more robust TFA assignment. I'm OK with role-based assignment, but something more robust than a permission. Can you select which TFA process to use by role (if there is more than one)?